File: CRM_project/Python/spotkanie.py
import datetime


class Spotkanie:
    __spotkania = []
    __nextID = 0
    def __init__(self):
        self.__id = Spotkanie.__nextID
        Spotkanie.__nextID += 1
        self.__konsultant = None
        self.__prowadzacy = None
        self.__klient_indywidualny = None
        self.__przedstawiciel = None
        self.__temat = ''
        self.__data = self.ustawDate()
        Spotkanie.__spotkania.append(self)
        print('Rozpoczyna się spotkanie')



    def dolacz(self, pracownikptr):
        if pracownikptr.getGroupId() == 0:
            self.__prowadzacy = pracownikptr
            string = "Spotkanie prowadzi " + self.__prowadzacy.getName()
            print(string)

        elif pracownikptr.getGroupId() == 1:        #jesli jest przedstawicielem klienta
            self.__przedstawiciel = pracownikptr
            string = 'Na spotkaniu ' + self.__przedstawiciel.getName() + ' jest przedstawicielem firmy ' + self.__przedstawiciel.getEmployer()
            print(string)

        elif pracownikptr.getGroupId() == 2:
            self.__konsultant = pracownikptr
            string = 'W spotkaniu bierze udział konsultant ' + self.__konsultant.getName()
            print(string)


    def dolaczIndywidualny(self, klientptr):
        self.__klient_indywidualny = klientptr
        print('W spotkaniu bierze udział klient indywidualny', klientptr.getName())


    def ustawDate(self):
        string = input('Kiedy odbywa się spotkanie? Data spotkania w formacie RRRR-MM-DD: ')
        year, month, day = map(int, string.split('-'))
        data = datetime.date(year, month, day)
        return data


    @staticmethod
    def znajdzOpId(id):
        for s in Spotkanie.__spotkania:
            if s.getId() == id:
                return s
        print('Brak spotkania o takim id')
        return None


    @staticmethod
    def wypiszIdSpotkan():
        print('Id odbytych spotkań:')
        for s in Spotkanie.__spotkania:
            print(s.getId())


    def getId(self):
        return self.__id

    
    def opisSpotkania(self):
        string = 'Id spotkania: ' + str(self.__id) + ' Data spotkania: ' + str(self.__data) + '\n'
        if self.__klient_indywidualny != None:
            string += 'Klient(indywidualny): ' + self.__klient_indywidualny.getName() + '\n'

        elif self.__przedstawiciel != None:
            string += 'Klient(firma): ' + self.__przedstawiciel.getEmployer() + ', w spotkaniu brał udział przedstawiciel ' + self.__przedstawiciel.getName() + '\n'

        string += 'Temat spotkania: ' + self.__temat + '\nSpotkanie prowadził/a ' + self.__prowadzacy.getName() + '\n'
        
        if self.__konsultant != None:
            string += 'W spotkaniu brał udział konsultant ' + self.__konsultant.getName() + '. Uwagi i spostrzeżenia konsultanta: \n' + self.__uwagi

        print(string)


    def dodajUwagi(self, uwagi_konsultanta):
        self.__uwagi = uwagi_konsultanta


    def dodajTemat(self, temat):
        self.__temat = temat
        print('Temat spotkania:', self.__temat)

File: CRM_project/Cpp/source/Spotkanie.cpp
#include "Spotkanie.hpp"

using namespace std;

Spotkanie::Spotkanie(){
    konsultant = nullptr;
    prowadzacy = nullptr;
    konsultant = nullptr;
    klient_indywidualny = nullptr;
    przedstawiciel = nullptr;
    _id = _nextId++;
    string input;
    cout << "Z kim odbywa się spotkanie? Wpisz imię i nazwisko klienta, bądź nazwę firmy.\n";
    getline(cin, input);
    Klient::spotkajSie(input, this);
    cout << "Jaki jest temat spotkania?\n";
    getline(cin, input);
    _temat = input;
    cout << "Który z Twoich pracowników prowadzi spotkanie z klientem?\n";
    getline(cin, input);
    NaszPracownik::prowadzSpotkanie(input, this);
    cout << "Czy w spotkaniu bierze udział konsultant? Wpisz 'tak' lub 'nie'.\n";
    getline(cin, input);
    if(input == "tak"){
        cout << "Podaj imię i nazwisko swojego konsultanta\n";
        getline(cin, input);
        Konsultant::konsultuj(input, this);
    }else{}

    Spotkanie::setDate();
    cout << "Spotkanie zapisane.\n\n";
    _spotkania.push_back(this);
}

Spotkanie::~Spotkanie(){
    for(size_t i=0; i<_spotkania.size(); i++){
        if(_spotkania[i] == this){
            _spotkania[i] = _spotkania.back();
            _spotkania.pop_back();
            break;
        }
    }
    cout << "Spotkanie usunięte\n";
}

void Spotkanie::setDate() {
    int day, month, year;
    int c = 0;
    cout << "Podaj datę w spotkania formacie DD/MM/RRRR\nData: ";
            cin >> day;
             if(cin.get() != '/'){
                cout << "Zły format, spodziewany slash: '/'\n";
                c++;
            }
            else{
                cin >> month;
                if(cin.get() != '/'){
                    cout << "Zły format, spodziewany '/'\n";
                    c++;
                }
                else{
                    cin >> year;
                    _day = day;
                    _month = month;
                    _year = year;
                    cout << "Data ustawiona pomyślnie\n";
                }
            }
            while(cin.fail()){
                cin.clear();
                cout << "Błąd(!): Data nie powinna zawierać liter. Usuwam spotkanie.\n";
            }
            if(c != 0){
                setDate();
            }
}

void Spotkanie::dolacz(Pracownik *ptr){
    if(ptr->getGroupId() == 1){     //jesli pracownik jest przedstawicielem klienta
        przedstawiciel = (PrzedstawicielKlienta*)ptr;
        cout << ptr->getName() << " jest przedstawicielem " << przedstawiciel->getEmployer() << " na spotkaniu.\n";
    }
    else if(ptr->getGroupId() == 0){
        prowadzacy = (NaszPracownik*)ptr;
        cout << "Spotkanie prowadzi " << prowadzacy->getName() << "\n";
    }
    else if(ptr->getGroupId() == 2){
        konsultant = (Konsultant*)ptr;
        cout << "W spotkaniu bierze udział konsultant " << konsultant->getName() << "\n";
    }
}

void Spotkanie::dolaczIndywidualny(Klient* ptr){
    klient_indywidualny = (Klient*)ptr;
    cout << "W spotkaniu weźmie udział klient indywidualny " << klient_indywidualny->getName() << "\n";
}

string Spotkanie::opisSpotkania(){
    string s = "Id spotkania: " + to_string(_id) + "  Data spotkania: " + to_string(_day) + "." + to_string(_month) + "." + to_string(_year) + "\n" + "Klient";
    if(klient_indywidualny != nullptr){
        s += "(indywidualny): " + klient_indywidualny->getName() + "\n";
    }
    else if(przedstawiciel != nullptr){
        s += "(firma): " + przedstawiciel->getEmployer() + ", w spotkaniu brał udział przedstawiciel " + przedstawiciel->getName() + "\n";
    }
    s += "Temat spotkania: " + _temat + "\nSpotkanie prowadził/a " + prowadzacy->getName() + "\n";
    if(konsultant != nullptr){
        s += "W spotkaniu brał udział konsultant "  + konsultant->getName() + "\nUwagi i spostrzeżenia konsultanta: " + _uwagi + "\n";
    }
    return s;
}

void Spotkanie::dodajUwagi(string uwagi){
    _uwagi = uwagi;
}

void Spotkanie::wypiszIdSpotkan(){
    for(auto s : _spotkania){
        cout << s->getId() << "\n";
    }
}

Spotkanie* Spotkanie::znajdzPoId(int id){
    for(auto s : _spotkania){
        if(s->getId() == id){
            return s;
            break;
        }
    }
    cout << "Brak spotkania o takim id\n";
    return nullptr;
}

int Spotkanie::getId(){ return _id; }
int Spotkanie::_nextId = 0;
vector<Spotkanie*> Spotkanie::_spotkania;